# CH47 에러 처리

## 47.1 에러 처리의 필요성

&nbsp;에러나 예외적인 상황에 대응하지 않으면 프로그램은 강제 종료될 것

## 47.2 try...catch...finally 문

```javascript
try {
  // 실행할 코드(에러가 발생할 가능성이 있는 코드)
} catch (err) {
  // try 코드 블록에서 에러가 발생하면 이 코드 블록의 코드가 실행된다.
  // err에는 try 코드 블록에서 발생한 Error 객체가 전달된다.
} finally {
  // 에러 발생과 상관없이 반드시 한 번 실행된다.
}
```

## 47.3 Error 객체

```javascript
const error = new Error("invalid");
```

- Error 생성자 함수가 생성한 에러 객체는 message 프로퍼티와 stack 프로퍼티를 갖음
- message 프로퍼티의 값은 Error 생성자 함수에 인수로 전달한 에러 메시지
- stack 프로퍼티의 값은 에러를 발생시킨 콜스택의 호출 정보를 나타내는 문자열이며 디버깅 목적으로 사용

## 47.4 throw 문

```javascript
// 에러 발생시키기
throw 표현식;
```

## 47.5 에러의 전파

&nbsp;에러는 호출자 방향으로 전파된다. 즉, 콜 스택 아래 방향(실행 중인 실행 컨텍스트가 푸시되기 직전에 푸시된 실행 컨텍스트 방향)으로 전파된다.
&nbsp;주의할 것은 비동기 함수인 setTimeout이나 프로미스 후속 처리 메서드의 콜백 함수는 호출자가 없다는 것이다.

-> 태스크 큐나 마이크로 태스크 큐에 일시 저장되었다가 콜 스택이 비면 이벤트 루프에 의해 콜 스택으로 푸시되어 실행됨 -> 앞에 아무도 없음 이슈
